src/common/src/array/arrow/arrow_deltalake.rs

```diff
@@ -0,0 +1,23 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use arrow_impl::to_record_batch_with_schema as to_deltalake_record_batch_with_schema;
+use {
+    arrow_array_deltalake as arrow_array, arrow_buffer_deltalake as arrow_buffer,
+    arrow_cast_deltalake as arrow_cast, arrow_schema_deltalake as arrow_schema,
+};
+
+#[allow(clippy::duplicate_mod)]
+#[path = "./arrow.rs"]
+mod arrow_impl;
```

Comment:
Wow, this is quite interesting! 😮 How did you come up with this? Any other projects used this trick?

> FYI, I tried forked delta-rs and upgrades their datafusion version, but it cannot compile without some extra effort because there is some breaking changes in the upgrade.

>> ... there will always be a time when icelack and delta.rs depend on an inconsistent arrow.
>> If we want to finally unify both of their dependencies on arrow, i.e. only one version of arrow exists in our system, I think we can totally use 48 in this pr...
> 
> Not only deltalake and icelake are using arrow, but also our internal udf. In the future if we upgrade the arrow version of udf, it doesn't make sense to be blocked by arrow version of some external connectors. I think it's acceptable to support several versions of arrow at the same time if we don't have to duplicate the code.

So my thoughts:
- Agree that upgrade effort might be non-trivial. So blocking everything before unifying deps version might be ineffective.
- At the beginning, I don't think we might need to upgrade solely the arrow version of udf. Therefore, using the greatest common divisor SGTM, but indeed it may break someday when we want to use a feature on the latest version of one dependency, and it already upgraded its arrow version. So we will run into this problem again. Then we have to consider how to live with multiple versions of arrow.
- I'm generally OK with multiple versions. Only concerns are: 
  1. maintainability. Current solution LGTM
  2. API surface: Will we ever need to e.g., pass arrow48 to arrow49?

Nits:
- `#[allow]` -> `#[expect]`
- prefer using version number as the suffix, instead of `common` (this is especially confusing to me. What does "common" mean?) and `deltalake`.

Changes requested:
- Add some comments on `arrow/mod.rs` and/or above `mod arrow_impl;` about how this trick works.
- Add some comments on `arrow_*` about where it is used. i.e., why do we have to pass arrow data around.

Reply:
> How did you come up with this? Any other projects used this trick?

Initially in https://github.com/risingwavelabs/risingwave/pull/13600/commits/c890167eb4dbee6bff7750e85740272a1c5a3781 I use macro too generate the same code for different version with the same trick, which set different arrow version to the same arrow mod name for different conversion code implementation. But I think this is too hacky and the IDE will fail to index a large part of code. And then I looked for approach to include the code within an outer file to a file so that I don't have to include the code in a long macro, and I found that the current approach is a perfect match😄 And I don't need any other project using this trick yet.

> Will we ever need to e.g., pass arrow48 to arrow49?

Currently arrow does not flow in our internal system. It's only used to interact with external systems, so in short term I don't think there will be feature requirement on arrow48 to arrow49.